src/@next/components/organisms/ProductGallery/ProductGallery.tsx

```diff
@@ -0,0 +1,124 @@
+
+  if (imageIndex >= images.length && imageIndex > 0) {
+    setImageIndex(0);
+  }
+
+  const bottomImageRef = React.useRef<HTMLDivElement | null>(null);
+  const topImageRef = React.useRef<HTMLDivElement | null>(null);
+  const [topImageIntersectionObserver, topImageInView] = useInView({
+    threshold: 0.5,
+  });
+
+  const [bottomImageIntersectionObserver, bottomImageInView] = useInView({
+    threshold: 0.5,
+  });
+
+  const setBottomRef = React.useCallback(
+    node => {
+      bottomImageRef.current = node;
+      bottomImageIntersectionObserver(node);
+    },
+    [bottomImageIntersectionObserver]
+  );
+
+  const setTopRef = React.useCallback(
+    node => {
+      topImageRef.current = node;
+      topImageIntersectionObserver(node);
+    },
+    [topImageIntersectionObserver]
+  );
+
+  const setIntersectionObserver = (index: number, lengthOfArray: number) => {
+    if (lengthOfArray > MINIMAL_NUMBER_OF_IMAGES_FOR_BUTTONS) {
+      if (index === 0) {
+        return setTopRef;
+      }
+      if (index === lengthOfArray - 1) {
+        return setBottomRef;
+      }
+    }
+  };
+
+  return (
+    <S.Wrapper>
+      <S.ThumnbanilsContainer>
+        {!topImageInView && displayButtons && (
+          <S.TopButton
+            onClick={() => {
+              if (topImageRef.current) {
+                topImageRef.current.scrollIntoView({
+                  behavior: "smooth",
+                  block: "end",
+                  inline: "nearest",
+                });
+              }
+            }}
+          >
+            <Icon name="select_arrow" size={10} />
+          </S.TopButton>
+        )}
+        {!bottomImageInView && displayButtons && (
+          <S.BottomButton
+            onClick={() => {
+              if (bottomImageRef.current) {
+                bottomImageRef.current.scrollIntoView({
+                  behavior: "smooth",
+                  block: "end",
+                  inline: "nearest",
+                });
+              }
+            }}
+          >
+            <Icon name="select_arrow" size={10} />
+          </S.BottomButton>
+        )}
+        <S.ThumbnailList>
+          <ul>
+            {images &&
+              images.length > 0 &&
+              images.map((image, index) => {
+                return (
+                  <li>
+                    <S.Thumbnail
+                      ref={setIntersectionObserver(index, images.length)}
+                      onClick={() => setImageIndex(index)}
+                      onMouseEnter={() => setImageIndex(index)}
+                      activeThumbnail={Boolean(index === imageIndex)}
+                    >
+                      <CachedImage alt={image.alt} url={image.url} />
+                    </S.Thumbnail>
+                  </li>
+                );
+              })}
+          </ul>
+        </S.ThumbnailList>
+      </S.ThumnbanilsContainer>
+
+      <S.Preview>
+        {images && images.length > 0 && imageIndex < images.length && (
```

Comment:
Here again, why is there needed check `imageIndex < images.length`?  I don't see call of `setImageIndex()` with index bigger then images length and we start with `useState(0)`

Reply:
If we pick image at the end of images array and then pick variant with lesser number of images we cat get array out of index error. I added useEffect call to reset state in case of images prop change.